Stream state updates from icinga:runtime:state

lippserd opened the pull request:
Icinga now sends runtime updates in two separate channels,
icinga:runtime for config updates and icinga:runtime:state for
state updates. With this change, Icinga DB reads from these two
streams. This is a preparation so that state updates can be
streamed directly after a (re)start of Icinga or Icinga DB without
waiting for the config sync, as it is currently done.

refs Icinga/icinga2#8915

julianbrost reviewed (changes requested):
Seems like `XRead().Result()` only returns the streams with messages. I added `r.logger.Infow("xread", zap.Strings("streams", streams))` right before the call to `XRead()` and this is what I've got:
```
2021-07-29T14:29:59.030+0200	INFO	icingadb/runtime_updates.go:102	xread	{"streams": ["icinga:runtime", "icinga:runtime:state", "1627561541341-9", "1627561591601-0"]}
2021-07-29T14:30:06.061+0200	INFO	icingadb/runtime_updates.go:102	xread	{"streams": ["icinga:runtime", "1627561806059-0"]}
2021-07-29T14:30:06.062+0200	INFO	icingadb/runtime_updates.go:102	xread	{"streams": ["icinga:runtime", "1627561806060-1"]}
```

So looks like the current code will only sync either of the streams now (unless you happen to have new messages in both streams each time).

Suggestion: make the last IDs a `map[string]string` from key to ID throughout the code (I find the []string{key1, key2, ..., id1, id2, ...)` a strange format anyways) and convert it to the required `[]string` for each `XRead()`. Other option would be to just update the IDs in-place in `streams`, but that would be some ugly indexing work.

lippserd commented:
> Seems like `XRead().Result()` only returns the streams with messages. I added `r.logger.Infow("xread", zap.Strings("streams", streams))` right before the call to `XRead()` and this is what I've got:
> 
> ```
> 2021-07-29T14:29:59.030+0200	INFO	icingadb/runtime_updates.go:102	xread	{"streams": ["icinga:runtime", "icinga:runtime:state", "1627561541341-9", "1627561591601-0"]}
> 2021-07-29T14:30:06.061+0200	INFO	icingadb/runtime_updates.go:102	xread	{"streams": ["icinga:runtime", "1627561806059-0"]}
> 2021-07-29T14:30:06.062+0200	INFO	icingadb/runtime_updates.go:102	xread	{"streams": ["icinga:runtime", "1627561806060-1"]}
> ```
> 
> So looks like the current code will only sync either of the streams now (unless you happen to have new messages in both streams each time).
> 
> Suggestion: make the last IDs a `map[string]string` from key to ID throughout the code (I find the `[]string{key1, key2, ..., id1, id2, ...)` a strange format anyways) and convert it to the required `[]string` for each `XRead()`. Other option would be to just update the IDs in-place in `streams`, but that would be some ugly indexing work.

Good catch. I introduced a new type for the streams:

https://github.com/Icinga/icingadb/pull/325/files#diff-dc1ad4347acc2a73e8d830a986f72b63a442ae0f678711ee3c91651d53cf17fbR16

julianbrost commented on pkg/icingaredis/utils.go:
```diff
@@ -12,6 +12,25 @@ import (
+	streams := make([]string, 0, len(s)*2)
+	ids := make([]string, 0, len(s))
+
+	for k, i := range s {
```
I find using `i` as a non-counter variable in a loop counterintuitive. Spelling out the variable(s) here won't hurt I guess.
